### DISCUSSION TITLE
DFU device cannot download data: endpoint stalled or request not supported.

### DISCUSSION THREAD

**Discussion by zack-lee-0126** [fwupd/fwupd#2687] [General]
Hi im here to ask a question about firmware update problem. im using F33 in desktop and trying to update our device firmware through our local, and the result showed as below. 
![Screenshot from 2020-12-14 17-07-33](https://user-images.githubusercontent.com/76022929/102173928-356a9400-3ed7-11eb-8678-a728bb629960.png)

**Comment by superm1** [comment]
What I would suggest to do is run `fwupdtool get-devices` and see if the device is missing.  If it's missing, check `fwupdtool get-devices --plugins=dfu --verbose` output to see why.

Then plug it back in and check if the device shows up.  If it does, please repeat the installation like this:

`fwupdtool install 1.cab --verbose --plugins=dfu`

So we can see what actions are followed.

**Reply by zack-lee-0126** [reply]
we did get our device but it show a information " 
![Screenshot from 2020-12-15 13-59-41](https://user-images.githubusercontent.com/76022929/102177614-e1fc4400-3ede-11eb-990f-20627dc7bb56.png)

and we continued the installation just like you mentioned then we get the information showed below.
![Screenshot from 2020-12-15 14-03-50](https://user-images.githubusercontent.com/76022929/102177803-38698280-3edf-11eb-99f7-15dc1d6e4ef6.png)
![Screenshot from 2020-12-15 14-03-55](https://user-images.githubusercontent.com/76022929/102177810-3bfd0980-3edf-11eb-8ead-0d6367b91ace.png)

**Reply by superm1** [reply]
The missing updatable flag was first noticed on Realtek camera that seemed to not follow DFU spec.  This keyboard might be in the same boat.

Is this an engineering sample really without a serial number?  It seems odd to me that no serial number listed, and I would worry if it failed to install properly.

Regarding final error message - I would ask is it expected that user needs to power cycle device manually?

**Reply by zack-lee-0126** [reply]
i will check the DFU spec as you mentioned, thank you. 

this is a product that had been mass manufactured and is a old sample that we tried to update on our own.

 we had already upload the CAB file to the LVFS for testing, but we still out of permission to push on to the testing mode. So we were trying to be end user and update the firmware from our CAB file, but it seems difficult as we meet a lot of problems.

so what suggestions you may give us?

**Reply by superm1** [reply]
> this is a product that had been mass manufactured and is a old sample that we tried to update on our own.

I'm sorry this is unclear.  My interpretation of your statement: so the product is now in mass manufacturing but this happens to be a pre-production sample you're using so the lack of a serial number may be expected?  Is that right?

>we had already upload the CAB file to the LVFS for testing, but we still out of permission to push on to the testing mode. So we were trying to be end user and update the firmware from our CAB file, but it seems difficult as we meet a lot of problems.

Yeah this is the right approach to do, make sure it works from the CAB file before it can go more widely to a customer's side.

>so what suggestions you may give us?

1. Can you please confirm with your colleagues that normally flash on Windows whether the keyboard is required to be unplugged/replugged when update completes?  If so, we probably need to set up a set of quirks to set the behavior here too.

2. Can you please get two firmware versions you can move interchangeably between and try to swap between them and confirm whether the version is actually changing or if the firmware flash process is entirely failing, and not just that it isn't coming back.

**Reply by zack-lee-0126** [reply]
> I'm sorry this is unclear. My interpretation of your statement: so the product is now in mass manufacturing but this happens to be a pre-production sample you're using so the lack of a serial number may be expected? Is that right?
> 
Yes, you are right.
> 
> Yeah this is the right approach to do, make sure it works from the CAB file before it can go more widely to a customer's side.
> 
> 
> 1. Can you please confirm with your colleagues that normally flash on Windows whether the keyboard is required to be unplugged/replugged when update completes?  If so, we probably need to set up a set of quirks to set the behavior here too.

yes. when we are doing normal flash on window there will be a unplugged and replugged, when the fwupd is done.

> 2. Can you please get two firmware versions you can move interchangeably between and try to swap between them and confirm whether the version is actually changing or if the firmware flash process is entirely failing, and not just that it isn't coming back.

the firmware is doing fine in window OS, the version did changed in WINDOWS but not in a FEDORA. 

so what should we do about the behavior of the replugged ?

**Reply by zack-lee-0126** [reply]
there is one more question on the unplug and replug, as you can see the information in red. Did the device restart successfully? from the way i see it, it did . How come it fail in the second time? need some of your opinions~ thank 
![Screenshot from 2020-12-15 00-30-08V2](https://user-images.githubusercontent.com/76022929/102291080-440b8680-3f7d-11eb-8253-bfa6075e4f62.png)
you

**Reply by superm1** [reply]
@hughsie From those answers I guess we pretty much need `FWUPD_DEVICE_FLAG_SKIPS_RESTART` to be applied to the device.  Is there anything else using DFU that needs this?  I didn't see anything else using it.

**Reply by hughsie** [reply]
@superm1 the reporter sent me a USB dump via email. I think I have a patch to fix it, but waiting on the reporter to verify.
